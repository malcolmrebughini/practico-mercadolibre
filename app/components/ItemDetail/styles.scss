@import 'app/styles/vars';


.container {
  background-color: #fff;
  margin: 0 auto;

  @media (min-width: 768px) {
    max-width: $body-width;
    border-radius: 4px;
  }

  .pic {
    & > img {
      width: 100%;
      height: auto;
    }

    @media (min-width: 768px) {
      width: 680px;
    }
  }

  .condition {
    margin-top: 32px;
    font-weight: 300;
    font-size: 14px;
    text-transform: capitalize;
  }

  .title {
    font-weight: bold;
    font-size: 24px;
    margin-top: 16px;
  }

  .infoContainer {
    padding: 0 32px;

    @media (min-width: 768px) {
      padding: 0;
      flex-shrink: 5;
      padding-left: 32px;
    }
  }

  .descriptionContainer {
    padding: 40px 32px;

    @media (min-width: 768px) {
      padding: 0;
      margin-top: 40px;
    }
  }

  .price {
    font-size: 46px;
    margin: 32px 0;

    .priceDecimals {
      font-size: 23px;
      vertical-align: top;
      line-height: 1.5;
    }
  }

  .buyButton {
    width: 100%;
    background: #3483fa;
    border: 1px solid #3483fa;
    font-size: 18px;
    color: #fff;
    border-radius: 4px;
    padding: 12px;
  }

  .descriptionTitle {
    font-size: 28px;
  }

  .description {
    color: #666;
    margin-top: 32px;
    font-size: 16px;

    @media (min-width: 768px) {
      width: 65%;
    }
  }

  .innerContainer {
    @media (min-width: 768px) {
      display: flex;
    }
  }

  .someContainer {
    @media (min-width: 768px) {
      padding: 32px;
    }
  }
}