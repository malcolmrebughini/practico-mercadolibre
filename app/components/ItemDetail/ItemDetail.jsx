import React from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scss';


function ItemDetail(props) {
  const { data } = props;

  return (
    <div className={styles.container} itemScope itemType="http://schema.org/Product">
      <div className={styles.someContainer}>
        <div className={styles.innerContainer}>
          <div className={styles.pic}>
            <img src={data.pictures[0].url} alt={data.title} />
          </div>
          <div className={styles.infoContainer}>
            <div className={styles.condition}>
              {data.condition} - {data.sold_quantity} vendidos
            </div>
            <h1 className={styles.title} itemProp="name">{ data.title }</h1>
            <div className={styles.price}>
              $ {data.price.integer.toLocaleString('es-AR')}
              { data.price.decimals ?
                <span className={styles.priceDecimals}>{data.price.decimals}</span>
                : null
              }
            </div>
            <button className={styles.buyButton}>Comprar</button>
          </div>
        </div>
        <div className={styles.descriptionContainer}>
          <h2 className={styles.descriptionTitle}>Descripci√≥n del producto</h2>
          <p
            className={styles.description}
            itemProp="description"
            dangerouslySetInnerHTML={
              { __html: data.description.text || data.description.plain_text }
            }
          />
        </div>
      </div>
    </div>
  );
}


ItemDetail.propTypes = {
  data: PropTypes.shape({
    pictures: PropTypes.array,
    price: PropTypes.shape({
      integer: PropTypes.number,
      decimals: PropTypes.number,
    }),
    description: PropTypes.shape({ plain_text: PropTypes.string }),
    condition: PropTypes.string,
    sold_quantity: PropTypes.number,
    title: PropTypes.string,
  }),
};


ItemDetail.defaultProps = {
  data: {},
};


export default ItemDetail;
