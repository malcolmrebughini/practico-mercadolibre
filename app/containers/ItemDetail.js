import { connect } from 'react-redux';
import ItemDetail from '../components/ItemDetail';


function mapStateToProps(state) {
  return {
    data: state.itemDetail.data,
  };
}


export default connect(mapStateToProps)(ItemDetail);
