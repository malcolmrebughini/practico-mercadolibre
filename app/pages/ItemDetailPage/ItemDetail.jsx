import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import TopBar from '../../containers/TopBar';
import Breadcrumbs from '../../containers/Breadcrumbs';
import ItemDetail from '../../containers/ItemDetail';


class ItemDetailPage extends React.Component {
  componentWillMount = () => {
    if (Object.keys(this.props.data).length === 0) this.props.getItem(this.props.itemId);
  }

  componentWillUnmount = () => {
    this.props.clearData();
  }

  render = () => {
    const categories = this.props.categories.reduce((string, value, index, array) => {
      if (array.length === index + 1) return `${string} ${value}.`;
      return `${string} ${value},`;
    }, '');
    const price = this.props.data.price;
    const description =
      `Cómpralo en Mercado Libre a $ ${price} - Encuentra más productos de ${categories}`;

    return (
      <div>
        <TopBar />
        { !this.props.isLoading &&
          <div>
            <Helmet>
              <title>{`${this.props.data.title} - $ ${price} en MercadoLibre`}</title>
              <meta name="description" content={description} />
            </Helmet>
            <Breadcrumbs />
            <ItemDetail />
          </div>
        }
      </div>
    );
  }
}


ItemDetailPage.propTypes = {
  data: PropTypes.shape({
    price: PropTypes.number,
    title: PropTypes.string,
  }),
  itemId: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
  isLoading: PropTypes.bool,
  getItem: PropTypes.func,
  clearData: PropTypes.func,
};

ItemDetailPage.defaultProps = {
  data: {},
};


export default ItemDetailPage;
