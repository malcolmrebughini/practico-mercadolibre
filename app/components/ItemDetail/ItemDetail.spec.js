/* eslint-env jasmine */
/* eslint react/jsx-filename-extension: 0 */

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ItemDetail from './ItemDetail';


describe('ItemDetail', () => {
  it('should render correctly', () => {
    const itemData = {
      pictures: [{ url: 'http://someurl.com/someimage.jpg' }],
      condition: 'new',
      sold_quantity: 245,
      title: 'This is some title',
      price: {
        integer: 10000,
        decimals: 99,
      },
      description: { plain_text: 'This is some description' },
    };

    const wrapper = shallow(
      <ItemDetail data={itemData} />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
