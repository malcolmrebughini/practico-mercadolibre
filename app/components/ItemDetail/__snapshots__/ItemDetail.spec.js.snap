// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetail should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="someContainer"
  >
    <div
      className="innerContainer"
    >
      <div
        className="pic"
      >
        <img
          alt="This is some title"
          src="http://someurl.com/someimage.jpg"
        />
      </div>
      <div
        className="infoContainer"
      >
        <div
          className="condition"
        >
          new
           - 
          245
           vendidos
        </div>
        <h1
          className="title"
        >
          This is some title
        </h1>
        <div
          className="price"
        >
          $ 
          10000
          <span
            className="priceDecimals"
          >
            99
          </span>
        </div>
        <button
          className="buyButton"
        >
          Comprar
        </button>
      </div>
    </div>
    <div
      className="infoContainer2"
    >
      <h2
        className="descriptionTitle"
      >
        Descripci√≥n del producto
      </h2>
      <p
        className="description"
        dangerouslySetInnerHTML={
          Object {
            "__html": "This is some description",
          }
        }
      />
    </div>
  </div>
</div>
`;
